edu-1036 | Develop for durable executions | Go dev guide (#2429)
temporalio/documentation: docs-src/go/chapter-durable-execution/non-deterministic-code-changes.md

---
id: non-deterministic-code-changes
title: Non-deterministic code changes
description: History Replay, sometimes also called Workflow Replay, is the mechanism that Temporal uses to reconstruct the state of a Workflow Execution. Temporal provides Durable Execution via this Replay Functionality.
sidebar_label: Durability through Replays
tags:
  - go sdk
  - developer-guide-doc-type
  - event history
  - replay
  - durable execution
---

> The most important thing to take away from the section is to make sure you have an application versioning plan whenever you are developing and maintaining a Temporal Application that will eventually deploy to a production environment.

Reviewer: General comment: This section shows up in the [Develop for Durable Executions section](https://temporal-documentation-git-durable-execution-go.preview.thundergun.io/dev-guide/go/durable-execution#why-replay-test) of the Dev Guide under the "Non-deterministic code changes" heading, which follows a list with two items. It's not immediately clear to which of those two items ("code changes" or "intrinsic non-determinstic logic") this heading pertains. In other words, the heading mixes words from both list items. "How code changes can affect determinism" might be a better heading.

Also, I would recommend covering the "intrinsic non-deterministic logic" case first. It's intuitive that, for example, using a random number in a conditional statement can lead to a different execution path with each invocation. Once you understand why that's a problem, it becomes easier to understand why a code change involving only deterministic statements could break compatibility with Workflow Executions that were execution at the time of deployment.

Author: Clarified the list to match headings, and did the ordering switch.

> We cover versioning APIs and versioning strategies in other parts of the developer's guide, this chapter set the stage to understand why and how to approach those strategies.

Reviewer 2: https://developers.google.com/style/person?hl=en

Did we start using `we` in the docs? I don't remember if we have exceptions to the rule for `We cover` like we do `We recommened` ?

> We cover versioning APIs and versioning strategies in other parts of the developer's guide, this chapter set the stage to understand why and how to approach those strategies.

Reviewer 2: ```suggestion
We cover versioning APIs and versioning strategies in other parts of the developer's guide, this chapter sets the stage to understand why and how to approach those strategies.
```

> We cover versioning APIs and versioning strategies in other parts of the developer's guide, this chapter set the stage to understand why and how to approach those strategies.

Author: ```suggestion
Versioning APIs and versioning strategies are covered in other parts of the developer's guide, this chapter set the stage to understand why and how to approach those strategies.
```

> We cover versioning APIs and versioning strategies in other parts of the developer's guide, this chapter sets the stage to understand why and how to approach those strategies.

Author: ```suggestion
Versioning APIs and versioning strategies are covered in other parts of the developer's guide, this chapter sets the stage to understand why and how to approach those strategies.
```

<!--TODO ^ update with links to those places -->

### The Event History

> Let's inspect the Event History of a recent Background Check Workflow using the `temporal workflow show` command:

Reviewer 2: ```suggestion
Inspect the Event History of a recent Background Check Workflow using the `temporal workflow show` command:
```

I think it is stronger if talking directly to the user.

```shell
temporal workflow show \
 --workflow-id backgroundcheck_workflow \
 --namespace backgroundcheck_namespace
```

> You will see output similar to this:

Author: ```suggestion
You should see output similar to this:
```

> You will see output similar to this:

Reviewer 2: Switching voice.

Reviewer 3: This is probably a result of pick and mixing some of the existing language from the 102 course — I tried to be on top of this, but a good thing to be aware of in the future for these.

   2  2023-10-25T20:28:03Z  WorkflowTaskScheduled
   3  2023-10-25T20:28:03Z  WorkflowTaskStarted
   4  2023-10-25T20:28:03Z  WorkflowTaskCompleted
   5  2023-10-25T20:28:03Z  ActivityTaskScheduled
   6  2023-10-25T20:28:03Z  ActivityTaskStarted
   7  2023-10-25T20:28:03Z  ActivityTaskCompleted
   8  2023-10-25T20:28:03Z  WorkflowTaskScheduled
   9  2023-10-25T20:28:03Z  WorkflowTaskStarted
  10  2023-10-25T20:28:03Z  WorkflowTaskCompleted
  11  2023-10-25T20:28:03Z  WorkflowExecutionCompleted

Result:
  Status: COMPLETED
  Output: ["pass"]
```

The preceding output shows eleven Events in the Event History ordered in a particular sequence.
All Events are created by the Temporal Server in response to either a request coming from a Temporal Client, or a [Command](/concepts/what-is-a-command) coming from the Worker.

Let's take a closer look:

- `WorkflowExecutionStarted`: This Event is created in response to the request to start the Workflow Execution.
- `WorkflowTaskScheduled`: This Event indicates a Workflow Task is in the Task Queue.
- `WorkflowTaskStarted`: This Event indicates that a Worker successfully polled the Task and started evaluating Workflow code.

> - `3. WorkflowTaskCompleted`: This Event indicates that the Worker suspended execution and made as much progress that it could.

Reviewer 2: ```suggestion
- `WorkflowTaskCompleted`: This Event indicates that the Worker suspended execution and made as much progress that it could.
```
